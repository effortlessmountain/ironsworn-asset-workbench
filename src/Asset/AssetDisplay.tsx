import React from "react";
import sanitize from "./sanitize";

import { scaleRatio, AssetScale } from "./assetScaling";
import { FontConfig, makeFontStyles } from "./fonts";
import { Icon } from "./asset";

const WriteIn = (props: { writeIn?: string }) => {
  return props.writeIn ? <div className="write-in">{props.writeIn}</div> : null;
};

const Description = (props: { description: string }) => {
  return props.description ? (
    <div
      className="description"
      dangerouslySetInnerHTML={{ __html: sanitize(props.description) }}
    ></div>
  ) : null;
};

interface Ability {
  name?: string;
  filled: boolean;
  text: string;
}
export const AbilityDisplay = (props: { ability: Ability }) => {
  const createAbilityName = (name) => {
    if (name) {
      return <span className="ability-name">{name}</span>;
    } else {
      return "";
    }
  };

  //TODO: investigate parsing markdown or custom markup instead of allowing em, ul, li
  let sanitizedText = sanitize(props.ability.text);

  return (
    <div className="ability">
      <i className={props.ability.filled ? "dot filled" : "dot unfilled"}></i>
      <div className="ability-description">
        {createAbilityName(props.ability.name)}
        <span
          className="ability-text"
          dangerouslySetInnerHTML={{ __html: sanitizedText }}
        ></span>
      </div>
    </div>
  );
};

const TrackValue = (props: { track: number; value: number; scale: string }) => {
  if (props.value === 0) {
    return (
      <div className="value" key="zed">
        0
      </div>
    );
  } else if (props.value <= props.track) {
    return <div className="value number">+{props.value}</div>;
  } else {
    return (
      <div className="value empty">
        <svg
          xmlns="http://www.w3.org/2000/svg"
          version="1.1"
          preserveAspectRatio="none"
          viewBox="0 0 100 100"
          height={95 * scaleRatio[props.scale]}
          width={113 * scaleRatio[props.scale]}
        >
          <line
            x1="0"
            y1="100"
            x2="100"
            y2="0"
            stroke="rgb(65,64,66)"
            style={{ strokeWidth: 3.5 }}
          />
        </svg>
      </div>
    );
  }
};

const TrackDisplay = (props: {
  track: string[] | number;
  scale: string;
  style;
}) => {
  if (!props.track) {
    return null;
  } else {
    let innerEntries = [];

    if (Array.isArray(props.track)) {
      innerEntries = props.track.map((entry, index) => {
        return (
          <div className="value text" key={index}>
            {entry}
          </div>
        );
      });
    } else {
      let trackLength = Math.max(5, props.track);
      for (let i = 0; i <= trackLength; i++) {
        innerEntries.push(
          <TrackValue
            track={props.track}
            value={i}
            scale={props.scale}
            key={i}
          ></TrackValue>
        );
      }
    }
    return (
      <div className="track" style={props.style}>
        {innerEntries}
      </div>
    );
  }
};

const IconDisplay = (props: { icon: Icon; scale: string }) => {
  if (props.icon != null) {
    return (
      <div
        className="icon"
        style={{
          backgroundImage: `url(${props.icon.dataUri})`,
          backgroundRepeat: "no-repeat",
          backgroundSize: "100%",
        }}
      ></div>
    );
  } else {
    return null;
  }
};

interface Asset {
  fonts?: FontConfig;
  type: string;
  icon?: Icon;
  name: string;
  writeIn?: string;
  writeIn2?: string;
  description: string;
  abilities: Ability[];
  track?: number | string[];
}

interface AssetProps {
  asset: Asset;
  scale: AssetScale;
}

export const AssetDisplay = (props: AssetProps) => {
  let asset = props.asset;
  let fonts = makeFontStyles(asset.fonts);

  //TODO: consolidate fonts into one global link. Less calls and may fix the race condition with preview/Download and fonts.
  return (
    <div className={`asset ${props.scale}`}>
      <link rel="stylesheet" href={fonts.googleFontUrl} />
      <div className="main-matter">
        <div className="top">
          <div className="type" style={fonts.type}>
            {asset.type}
          </div>
          <IconDisplay icon={asset.icon} scale={props.scale} />
          <div className="asset-name" style={fonts.assetName}>
            {asset.name}
          </div>
        </div>
        <div className="details" style={fonts.details}>
          <WriteIn writeIn={asset.writeIn}></WriteIn>
          <WriteIn writeIn={asset.writeIn2}></WriteIn>
          <Description description={asset.description} />
          <div className="abilities">
            {asset.abilities.map((ability, index) => (
              <AbilityDisplay ability={ability} key={index}></AbilityDisplay>
            ))}
          </div>
        </div>
      </div>
      <TrackDisplay
        track={asset.track}
        scale={props.scale}
        style={fonts.track}
      />
    </div>
  );
};
