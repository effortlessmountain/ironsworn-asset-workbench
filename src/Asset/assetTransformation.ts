import { cloneDeep } from "lodash";
import {
  AssetDocumentV1,
  AssetDocumentV2,
  AssetDocument,
  UnspecifiedAssetDocument,
  SvgProperties,
  SvgIcon,
  createDataUri,
} from "./asset";

const extractPropertyValue = (key, svgString) => {
  const regexp = new RegExp(`${key}=(?:"|')(.*?)(?:"|')`);
  return (svgString.match(regexp) || [])[1];
  //TODO: intelligent error here?
};

export function transformSvgString(svgString: string): SvgProperties {
  return {
    d: extractPropertyValue("d", svgString),
    fill: extractPropertyValue("fill", svgString),
    fillOpacity: extractPropertyValue("opacity", svgString),
    viewBox: extractPropertyValue("viewBox", svgString),
  };
}

export function transformToV2(v1: AssetDocumentV1): AssetDocumentV2 {
  const v2: AssetDocumentV2 = {
    documentFormatVersion: 2,
    type: v1.type,
    name: v1.name,
    writeIn: v1.writeIn,
    track: v1.track,
    description: v1.description,
    abilities: v1.abilities,
  };
  if (typeof v1.fonts === "object") {
    v2.fonts = {
      assetTypeFontSize: v1.fonts.assetTypeFontSize,
      assetTypeFont: v1.fonts.assetTypeFont,
      assetNameFontSize: v1.fonts.assetNameFontSize,
      assetNameFont: v1.fonts.assetNameFont,
      detailsFontSize: v1.fonts.detailsFontSize,
      detailsFont: v1.fonts.detailsFont,
      trackFontSize: v1.fonts.trackFontSize,
      trackFont: v1.fonts.trackFont,
    };
  }
  if (typeof v1.icon === "string") {
    v2.icon = v1.icon;
  } else if (typeof v1.icon === "object") {
    v2.icon = {
      type: "svg",
      author: v1.icon.author,
      name: v1.icon.name,
      svg: transformSvgString(v1.icon.svg),
    };
  }
  return v2;
}

function transformSvgPropertiesToDataUri(
  properties: SvgProperties = { d: "", fill: "", fillOpacity: "", viewBox: "" }
): string {
  const svg = `<svg style="height: 512px; width: 512px;" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><g><path d="${properties.d}" fill="${properties.fill}" fill-opacity="${properties.fillOpacity}" ></path></g></svg>`;
  return createDataUri(svg);
}

export function transformToV3(v2: AssetDocumentV2): AssetDocument {
  let v3 = (cloneDeep(v2) as unknown) as AssetDocument;
  v3.documentFormatVersion = 3;

  if (v2.icon) {
    const v2Icon = v2.icon as SvgIcon;
    v3.icon = {
      name: v2Icon.name,
      author: v2Icon.author,
      dataUri: transformSvgPropertiesToDataUri(v2Icon.svg),
    };
  }

  return v3;
}

export function transformToLatest(
  asset: UnspecifiedAssetDocument
): AssetDocument {
  if (!asset.documentFormatVersion) {
    return transformToLatest(transformToV2(asset as AssetDocumentV1));
  } else if (asset.documentFormatVersion === 2) {
    return transformToV3(asset as AssetDocumentV2);
  } else {
    return asset as AssetDocument;
  }
}
