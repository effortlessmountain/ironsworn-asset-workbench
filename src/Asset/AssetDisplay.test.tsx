import React from "react";
import { AssetDisplay } from "./AssetDisplay";
import renderer from "react-test-renderer";
import { lightbearer, caveLion, ironclad } from "./exampleAssets";
import { transformToLatest } from "./assetTransformation";

[lightbearer, caveLion, ironclad].forEach((asset) => {
  it("renders the asset " + asset.name, () => {
    const assetAsLatestVersion = transformToLatest(asset);

    const tree = renderer
      .create(
        <AssetDisplay asset={assetAsLatestVersion} scale="full"></AssetDisplay>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
