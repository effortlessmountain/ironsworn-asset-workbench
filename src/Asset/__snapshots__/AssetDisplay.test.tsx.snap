// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the asset Cave Lion 1`] = `
<div
  className="asset full"
>
  <link
    href="https://fonts.googleapis.com/css2?family=Simonetta&family=PT+Serif&display=swap"
    rel="stylesheet"
  />
  <div
    className="main-matter"
  >
    <div
      className="top"
    >
      <div
        className="type"
        style={
          Object {
            "fontFamily": "\\"Simonetta\\"",
            "fontSize": "1.03em",
          }
        }
      >
        companion
      </div>
      <div
        className="icon"
        style={
          Object {
            "backgroundImage": "url(data:image/svg+xml;base64,PHN2ZyBzdHlsZT0iaGVpZ2h0OiA1MTJweDsgd2lkdGg6IDUxMnB4OyIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2aWV3Qm94PSIwIDAgNTEyIDUxMiI+PGc+PHBhdGggZD0iTTExMC4wNTYgNjQuODE1Yy00LjIzNC4wMjctOC4zNTUuNTg3LTEyLjMzNyAxLjc5OUM4My4xMyA3MS4wNTQgNzIuOTMgNzcuMDMgNjUuMjQgODcuMzMzYy03LjY5MSAxMC4zMDMtMTMuMTIyIDI1LjcxNy0xNS41MTYgNDkuNzEzLS42NjkgNi43MDggMi4wMTIgMTguMzg0IDcuNzUgMzAuOTg2IDUuNzM4IDEyLjYwMiAxNC4yNDggMjYuMjc2IDIzLjgyOSAzOC4zODdsMi40OSAzLjE0Ni0uNjc4IDMuOTU1Yy0zLjA5NyAxOC4wOTEtMy42NDQgNTAuNzA2LS4yNTIgNjguNzc4LjgzNiA0LjQ1NCA1Ljc2NiAxNC40OTcgMTMuNjExIDI2LjI5NiA3Ljg0NiAxMS44IDE4LjQxMiAyNS44MjkgMzAuMDMyIDQwLjk0NCAyMy4yMzkgMzAuMjMgNTAuNjc4IDY0Ljc3MyA2OS44MSA5Ni41NDcuMDI0LjAzOC4zNDcuNTI1IDIuMTM5Ljg4NiAxLjc5MS4zNjIgNC42MDUuMjggNy4xMjctLjM1IDIuNDQxLS42MDkgNC41MjQtMS43NjYgNS4zMzgtMi40NTgtOS45NDYtNDIuMjg2LTE2LjE0LTg0LjE4NS0zNy41MS0xMjUuMTRsLTEwLjEwMi0xOS4zNTggMjAuODA5IDYuNjE3YzIxLjY0IDYuODgyIDQ4LjcxOCA4Ljg5NyA3My4zOTYgNy43MTkgMjQuNjc5LTEuMTc5IDQ3LjI2LTUuNjgxIDU4Ljg2OC0xMC4zNjJsNS42NDgtMi4yNzcgNC4yMTUgNC4zOTdjNDMuMjI1IDQ1LjA4OSA3My40MjcgOTguMDQ4IDExMi42NDQgMTQwLjkzNS4xMi4xMzItLjEzOS4xNi45MS4wODggMS4wNS0uMDcyIDMtLjcxNCA0LjcwNi0xLjkxNCAxLjcwNS0xLjIgMy4xMDYtMi44ODIgMy43MjItNC4yMzQuNjE3LTEuMzUzLjY1OS0yLjA4Ni4yNjgtMy4xNDktMjAuMzk2LTU1LjQ4Ny0zMC41NjUtMTA5Ljg5NC02MS44NC0xNjIuMjU4bC0yLjEwMy0zLjUyMSAxLjI3OS0zLjljNi44NjktMjAuOTIzIDE3Ljg1Mi00Mi43NjggNDIuOTAyLTYwLjMwM2w0LjUzNS0zLjE3NiA0Ljg4MSAyLjYxN2MxMS40NTQgNi4xNDUgMTkuMTIzIDUuODczIDI1LjY5IDEuOTM2IDEzLjEzMi03Ljg3NSAyNC4xNS0yMy44OCAyOC42MjUtMzcuNzg0bC0xNC43NjgtOC4zMS0uOTQzLTMuNzQ2Yy00LjAyNS0xNS45Ni04LjI1NS0yNS40OC0xNC42MjUtMzIuNzE5LTYuMzctNy4yNC0xNS41NzYtMTIuODk2LTMwLjk3Ny0xOS41NDNsLTYuNDI2LTIuNzczIDEuMTA0LTYuOTFjLjg5LTUuNTcuMDUyLTExLjQ4MS0xLjE4NC0xNy43NjgtMzYuMjI4IDI5LjkxNS01Mi44OTIgNjQuNC04Ni41MzMgMTAwLjA4NmwtMy4xOTcgMy4zOS00LjYxMy0uNjUyYy01Ny44NC04LjE4Ny0xNDIuOTI2LTUuMjU3LTE5OC40NTUgMS40MzYtNi4wNDUuNzI4LTExLjI2Mi0xLjE0LTE2LjM0LTMuNzMzLTUuMDc4LTIuNTkzLTkuOTk2LTYuMi0xNC40NzUtMTAuNDkyLTguOTU3LTguNTg0LTE2LjkzOC0xOS44Ny0xNi4wMS0zMy4yNTguNjU0LTkuNDI2IDIuNjgzLTE3LjA0OCA3LjAwNi0yMy4wMSA0LjMyMy01Ljk2MSAxMS4xMy05LjQzNCAxNy44MTUtMTAuMjcgMTMuMzY4LTEuNjY5IDI2LjkxNiAzLjgzMyA0Mi4wNTggMTAuMDQgMzAuMjg1IDEyLjQxNSA2NS45ODcgMjkuMDEgOTMuMDIyIDIxLjcxNyAxMC41NTQtMi44NDggMzAuNjc0LTIwLjA4OSA0NC40LTM3LjgzNCA2Ljg2My04Ljg3MyAxMi40NDItMTcuODYxIDE1LjM3Ny0yNC41MjggMS4zMzktMy4wNDEgMi4wMTMtNS40ODcgMi4yNC02LjkyMy0uNTA3LS4xMDYtMS4xNTMtLjE2My0yLjM1NS0uMDgtMi42NjUuMTgyLTcuMDY5IDEuMzc1LTEyLjU2IDMuODE0LTEwLjk4NSA0Ljg3Ny0yNi4yMzcgMTQuNDIzLTQ0LjY3NSAyNi44OTYtMTEuOTg0IDguMTA4LTI2LjAxNyA4LjM4Ni0zOC45OTIgNS4xNTMtMTIuOTc1LTMuMjMzLTI1Ljc0My05LjYyNi0zOC41OC0xNi4wNC0xOS4yNTUtOS42MTktMzguMjEzLTE4LjgxLTU1LjQ1LTIwLjA3N2E1My42NTcgNTMuNjU3IDAgMCAwLTQuMjctLjE0N3ptMzQzLjA5NiA3My4wMDZsOS44MzIgMjIuMTIzLTIyLjQ3NC0zLjUxMiAxMi42NDItMTguNjExek03NS4yOCAzMDguODU4Yy0xNC4yMzUgNDkuMDI4LTM1LjU0OCA5Ny4wNzItNTUuNjMzIDEzMi45NDctLjAxNC4wMjYtLjAyLjAxNS0uMDMuMDI4YS42NTEuNjUxIDAgMCAxIC4xMjIuMjA3Yy4yNTguNTkyIDEuMjk1IDEuOTYxIDIuODU2IDMuMDM1IDEuNTYgMS4wNzQgMy41NTMgMS44MjggNC45NDUgMi4wMDggMS4zOTEuMTggMS44MzYtLjExIDEuODQtLjExNCAzMy42NjktMjQuOTE0IDU4LjcxNy01NS40IDgzLjQ3LTg1LjY2OGwtLjYxMy0uNzk0Yy0xMS42ODItMTUuMTk3LTIyLjQyNy0yOS40MjgtMzAuNzUtNDEuOTQ2LTIuMjUtMy4zODQtNC4zMjEtNi42LTYuMjA3LTkuNzAzek0xOS42MTUgNDQxLjgzM2MtLjEwOC0uMTE0LS4wOTkuMTEgMCAwem0yOTcuMzc3LTExOS40ODVjLTcuMTI5IDIuMzY2LTE1LjUzMyA0LjQwOC0yNC44NjMgNi4wMjItMTIuNDE4IDQwLjA5MS0zMy4zMjcgNzcuODY3LTUyLjIwMyAxMTIuNjA1bC0uMTQxLjI2LS4xNTguMjVjLS4zMS40OTEtLjI1Mi4xODYtLjA0OS43ODNzMS4wMTIgMS44MSAyLjI3IDIuNzA3YzIuNTE0IDEuNzk2IDQuODIgMi4zNzQgNy4yMDguMjNsLjA0Ny0uMDQuMDQ3LS4wNDFjMzIuMjg1LTI4LjE3IDU5LjI1NC02MS40NTggODUuMjQyLTEwMi44NjctNS42MjEtNi43OTktMTEuMzk2LTEzLjQ1NS0xNy40LTE5LjkwOXoiIGZpbGw9IiNmZmYiIGZpbGwtb3BhY2l0eT0iMSIgPjwvcGF0aD48L2c+PC9zdmc+)",
            "backgroundRepeat": "no-repeat",
            "backgroundSize": "100%",
          }
        }
      />
      <div
        className="asset-name"
        style={
          Object {
            "fontFamily": "\\"Simonetta\\"",
            "fontSize": "1.26em",
          }
        }
      >
        Cave Lion
      </div>
    </div>
    <div
      className="details"
      style={
        Object {
          "fontFamily": "\\"PT Serif\\"",
          "fontSize": "0.97em",
        }
      }
    >
      <div
        className="write-in"
      >
        Name
      </div>
      <div
        className="description"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Your cat takes down its prey.",
          }
        }
      />
      <div
        className="abilities"
      >
        <div
          className="ability"
        >
          <i
            className="dot unfilled"
          />
          <div
            className="ability-description"
          >
            <span
              className="ability-name"
            >
              Eager
            </span>
            <span
              className="ability-text"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "When your cat chases down big game, you may <em>Resupply</em> with +edge (instead of +wits). If you do, take +1 supply or +1 momentum on a strong hit.",
                }
              }
            />
          </div>
        </div>
        <div
          className="ability"
        >
          <i
            className="dot unfilled"
          />
          <div
            className="ability-description"
          >
            <span
              className="ability-name"
            >
              Inescapable
            </span>
            <span
              className="ability-text"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "When you <em>Enter the Fray</em> or <em>Strike</em> by sending your cat to attack, roll +edge. On a hit, take +2 momentum.",
                }
              }
            />
          </div>
        </div>
        <div
          className="ability"
        >
          <i
            className="dot unfilled"
          />
          <div
            className="ability-description"
          >
            <span
              className="ability-name"
            >
              Protective
            </span>
            <span
              className="ability-text"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "When you <em>Make Camp</em>, your cat is alert to trouble. If you or an ally choose to relax, take +1 spirit. If you focus, take +1 momentum.",
                }
              }
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="track"
    style={
      Object {
        "fontFamily": "\\"Simonetta\\"",
        "fontSize": "1.42em",
      }
    }
  >
    <div
      className="value"
    >
      0
    </div>
    <div
      className="value number"
    >
      +
      1
    </div>
    <div
      className="value number"
    >
      +
      2
    </div>
    <div
      className="value number"
    >
      +
      3
    </div>
    <div
      className="value number"
    >
      +
      4
    </div>
    <div
      className="value empty"
    >
      <svg
        height={95}
        preserveAspectRatio="none"
        version="1.1"
        viewBox="0 0 100 100"
        width={113}
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          stroke="rgb(65,64,66)"
          style={
            Object {
              "strokeWidth": 3.5,
            }
          }
          x1="0"
          x2="100"
          y1="100"
          y2="0"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`renders the asset Ironclad 1`] = `
<div
  className="asset full"
>
  <link
    href="https://fonts.googleapis.com/css2?family=Noto+Sans+JP&family=Russo+One&family=Quicksand&family=Krona+One&display=swap"
    rel="stylesheet"
  />
  <div
    className="main-matter"
  >
    <div
      className="top"
    >
      <div
        className="type"
        style={
          Object {
            "fontFamily": "\\"Noto Sans JP\\"",
            "fontSize": "1.03em",
          }
        }
      >
        combat talent
      </div>
      <div
        className="icon"
        style={
          Object {
            "backgroundImage": "url(data:image/svg+xml;base64,PHN2ZyBzdHlsZT0iaGVpZ2h0OiA1MTJweDsgd2lkdGg6IDUxMnB4OyIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2aWV3Qm94PSIwIDAgNTEyIDUxMiI+PGc+PHBhdGggZD0iTTI0Ni40ODggMjkuNWwtMTMuMjQ0IDM5LjcyNSAxNC4zODEgODYuMjc1aDE2Ljc1bDE0LjM4LTg2LjI3NUwyNjUuNTE1IDI5LjVoLTE5LjAyNnpNMTk1IDYxLjVsLTQwIDMwaDYzLjcwOWwtMy45NTMtMjMuNzI1IDIuMDkyLTYuMjc1SDE5NXptMTAwLjE1MiAwbDIuMDkyIDYuMjc1LTMuOTUzIDIzLjcyNUgzNTdsLTQwLTMwaC0yMS44NDh6TTY5LjQyIDkwLjcyN0w0MS42MzkgMjU3LjQxIDcxIDI4Ni43NzNWMTYyLjM3NWwyMC44NjMtNDEuNzI1TDY5LjQyIDkwLjcyN3ptMzczLjE2IDBsLTIyLjQ0MyAyOS45MjNMNDQxIDE2Mi4zNzV2MTI0LjM5OGwyOS4zNjEtMjkuMzYzLTI3Ljc4LTE2Ni42ODN6TTExNy41NjMgMTA5LjVMODkgMTY2LjYyNVY0MDBsMTEwIDgyLjV2LTg3LjkwNmwtODAtMzJWMjExLjU3NmwzNS4zOTMtNzIuMDc2aDcyLjMxNmwtNS0zMEgxMTcuNTYyem0xNzIuNzI4IDBsLTUgMzBoNzIuMjcyTDM5MyAyMTAuMzc1djE1Mi4yMTlsLTgwIDMyVjQ4Mi41TDQyMyA0MDBWMTY2LjYyNUwzOTQuNDM3IDEwOS41SDI5MC4yOTF6bS0xMjQuNjg0IDQ4TDEzNyAyMTUuNzU4djMuNzQyaDIzOHYtNC44NzVMMzQ2LjQzNyAxNTcuNWgtNjQuMTQ2bC0yLjY2NiAxNmgtNDcuMjVsLTIuNjY2LTE2aC02NC4xMDJ6TTEzNyAyMzcuNXY0NmgzMHYtNDZoLTMwem05NiAwdjQ2aDQ2di00NmgtNDZ6bTExMiAwdjQ2aDMwdi00NmgtMzB6bS0yMDggNjR2NDguOTA2bDQ2LjY3OCAxOC42NyAyNy4wMjktNjcuNTc2SDEzN3ptOTMuMDkyIDBsLTI5LjcwNSA3NC4yNjJMMjE3IDM4Mi40MDZ2NTEuMDY3bDM5IDkuNzUgMzktOS43NXYtNTEuMDY3bDE2LjYxMy02LjY0NC0yOS43MDUtNzQuMjYyaC01MS44MTZ6bTcxLjIwMSAwbDI3LjAzIDY3LjU3NkwzNzUgMzUwLjQwNlYzMDEuNWgtNzMuNzA3eiIgZmlsbD0iI2ZmZiIgZmlsbC1vcGFjaXR5PSIxIiA+PC9wYXRoPjwvZz48L3N2Zz4=)",
            "backgroundRepeat": "no-repeat",
            "backgroundSize": "100%",
          }
        }
      />
      <div
        className="asset-name"
        style={
          Object {
            "fontFamily": "\\"Russo One\\"",
            "fontSize": "1.76em",
          }
        }
      >
        Ironclad
      </div>
    </div>
    <div
      className="details"
      style={
        Object {
          "fontFamily": "\\"Quicksand\\"",
          "fontSize": "0.97em",
        }
      }
    >
      <div
        className="description"
        dangerouslySetInnerHTML={
          Object {
            "__html": "If you wear high-tech armor...",
          }
        }
      />
      <div
        className="abilities"
      >
        <div
          className="ability"
        >
          <i
            className="dot filled"
          />
          <div
            className="ability-description"
          >
            
            <span
              className="ability-text"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "When you equip or adjust your armor, choose one. <li>Lightly armored: When you <em>Endure Harm</em> in a fight, add +1 and take +1 momentum on a hit.</li> <li>Geared for war: Mark encumbered. When you <em>Endure Harm</em> in a fight, add +2 and take +1 momentum on a hit.</li>",
                }
              }
            />
          </div>
        </div>
        <div
          className="ability"
        >
          <i
            className="dot unfilled"
          />
          <div
            className="ability-description"
          >
            
            <span
              className="ability-text"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "When you <em>Clash</em> while you are geared for war, add +1.",
                }
              }
            />
          </div>
        </div>
        <div
          className="ability"
        >
          <i
            className="dot unfilled"
          />
          <div
            className="ability-description"
          >
            
            <span
              className="ability-text"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "When you <em>Compel</em> in a situation where strength of arms is a factor, add +2.",
                }
              }
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="track"
    style={
      Object {
        "fontFamily": "\\"Krona One\\"",
        "fontSize": "1.12em",
      }
    }
  >
    <div
      className="value text"
    >
      LIGHTLY ARMORED
    </div>
    <div
      className="value text"
    >
      GEARED FOR WAR
    </div>
  </div>
</div>
`;

exports[`renders the asset Lightbearer 1`] = `
<div
  className="asset full"
>
  <link
    href="https://fonts.googleapis.com/css2?family=Simonetta&family=PT+Serif&display=swap"
    rel="stylesheet"
  />
  <div
    className="main-matter"
  >
    <div
      className="top"
    >
      <div
        className="type"
        style={
          Object {
            "fontFamily": "\\"Simonetta\\"",
            "fontSize": "1.03em",
          }
        }
      >
        ritual
      </div>
      <div
        className="icon"
        style={
          Object {
            "backgroundImage": "url(data:image/svg+xml;base64,PHN2ZyBzdHlsZT0iaGVpZ2h0OiA1MTJweDsgd2lkdGg6IDUxMnB4OyIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2aWV3Qm94PSIwIDAgNTEyIDUxMiI+PGc+PHBhdGggZD0iTTU0LjE4NCAxOS41MTJsMTE2LjAyIDE0NC44MzhjLTExLjgwNiA5Ljc0LTIxLjQgMjIuMDY2LTI3LjkyMyAzNi4wOTVMMjAuNDkzIDE0Mi43M3Y2Mi45OTNsMTEzLjk4MyAxOS4xNzJjLTEuMTIgNi4wNTYtMS43MDcgMTIuMjk1LTEuNzA3IDE4LjY3IDAgMy41MDIuMTc3IDYuOTY1LjUyIDEwLjM3OGwtNzIuMzc2IDI3LjcgNzkuNDY3LjY4N2M0LjE2NSAxMC4xNzMgOS45MTcgMTkuNTM0IDE2Ljk0OCAyNy43NzVMMjAuNDkyIDQ0MS43djUyLjQ4NGg0Ny4zMzZsMTA1Ljg3LTE2OC42NWMxMS42NDQgOC43OTUgMjUuMjEyIDE1LjE2NSAzOS45NDUgMTguMzY2bC0yMy43IDE1MC4yODRIMjk0LjIyTDI1My4yNSAzNDQuNjljOC42NjgtMS41MjQgMTYuOTY0LTQuMTQgMjQuNzQ0LTcuN2wzNy4yNjIgNDIuMjY4LTIxLjMyNi01MS4zNjVjNi44ODUtNC43OTIgMTMuMTYtMTAuNDAyIDE4LjY4LTE2LjY5MmwxODIuNTMzIDE0Ni42MVYzNDguNTU2TDMyOS40MyAyODQuODY1YzQuMjktOS43MjIgNy4xMTgtMjAuMjI3IDguMTk1LTMxLjIzOGw4NC43LTI1LjkzLTg2Ljc4LTcuMDA4Yy0yLjI2LTkuODktNS45NTUtMTkuMjQyLTEwLjg1Mi0yNy44MjVsMTcwLjQ1Mi04Ny4zOThWMjMuOTI0TDMxMS45MTggMTc1LjE1Yy01LjMzLTUuOTUtMTEuMzQyLTExLjI4LTE3LjkxMi0xNS44NmwxNS41MzMtMzguMzgyLTI5Ljk2OCAyOS45NjdjLTkuNjA4LTQuNTk1LTIwLjAzNC03Ljc0LTMwLjk5Ni05LjE0OEwyNjkuODkgMTkuNTEyaC00MC4xMmwxLjI3IDEyMS40NzhjLTE5LjAxNy44MDgtMzYuNzE2IDYuODE2LTUxLjcwNiAxNi42NEw5NC4xMDIgMTkuNTFINTQuMTg0eiIgZmlsbD0iI2ZmZiIgZmlsbC1vcGFjaXR5PSIxIiA+PC9wYXRoPjwvZz48L3N2Zz4=)",
            "backgroundRepeat": "no-repeat",
            "backgroundSize": "100%",
          }
        }
      />
      <div
        className="asset-name"
        style={
          Object {
            "fontFamily": "\\"Simonetta\\"",
            "fontSize": "1.26em",
          }
        }
      >
        Lightbearer
      </div>
    </div>
    <div
      className="details"
      style={
        Object {
          "fontFamily": "\\"PT Serif\\"",
          "fontSize": "0.97em",
        }
      }
    >
      <div
        className="abilities"
      >
        <div
          className="ability"
        >
          <i
            className="dot filled"
          />
          <div
            className="ability-description"
          >
            
            <span
              className="ability-text"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "When you focus on a source of light and capture its essence, roll +wits. On a strong hit, set your light track to +6. On a weak hit, make it +3. Then, when you make a move to overcome or navigate darkness, you may add +2 and suffer -1 light.",
                }
              }
            />
          </div>
        </div>
        <div
          className="ability"
        >
          <i
            className="dot unfilled"
          />
          <div
            className="ability-description"
          >
            
            <span
              className="ability-text"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "You may use your light to <em>Strike</em> or <em>Clash</em> against a dark-dwelling foe. Choose the amount of light to unleash, and roll +light (instead of +iron or +edge). Suffer -light equal to that amount. On a hit, your harm is 1+your unleashed light.",
                }
              }
            />
          </div>
        </div>
        <div
          className="ability"
        >
          <i
            className="dot unfilled"
          />
          <div
            className="ability-description"
          >
            
            <span
              className="ability-text"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "When you perform this ritual, add +1 and take +1 momentum on a hit.",
                }
              }
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="track"
    style={
      Object {
        "fontFamily": "\\"Simonetta\\"",
        "fontSize": "1.42em",
      }
    }
  >
    <div
      className="value"
    >
      0
    </div>
    <div
      className="value number"
    >
      +
      1
    </div>
    <div
      className="value number"
    >
      +
      2
    </div>
    <div
      className="value number"
    >
      +
      3
    </div>
    <div
      className="value number"
    >
      +
      4
    </div>
    <div
      className="value number"
    >
      +
      5
    </div>
    <div
      className="value number"
    >
      +
      6
    </div>
  </div>
</div>
`;
